utils/pdfgeneration.js

```diff
@@ -0,0 +1,157 @@
+import {
+  getSolarProjectById,
+  updateSubscriberBill,
+  getSubscriberBillsByIds,
+  getOwnerById
+} from '../airtable/request';
+import Constants from '../Constants';
+import sendEmail from './email';
+import EmailGenerators from './emailCopy';
+import saveChartToFile from './charts/charts';
+import {
+  generateGenerationDataChart,
+  generateCostOverTimeChart
+} from './chartGeneration';
+import BillTemplate from './pdf/BillTemplate';
+
+dotenv.config();
+const { billSuccess, pdfSuccess } = EmailGenerators;
+
+// Get latest subscriber data
+const getLatestSubscriberData = async subscriberId => {
+  console.log(`Get latest data for Subscriber with ID: ${subscriberId}`);
+  const subscriber = await getOwnerById(subscriberId);
+  console.log('Getting solar project data...');
+  const solarProject = await getSolarProjectById(subscriber.solarProjectId);
+
+  if (!subscriber.subscriberBillIds) {
+    throw new Error(
+      `Could not find Bill for subscriber with ID ${subscriberId}`
+    );
+  }
+
+  console.log(
+    `Getting subscriber bill data for IDs: ${subscriber.subscriberBillIds}`
+  );
+  const bills = await getSubscriberBillsByIds(subscriber.subscriberBillIds);
+  const latestBill = bills[bills.length - 1];
+  const previousBills = bills.slice(0, bills.length - 1);
+  return { subscriber, solarProject, latestBill, previousBills };
+};
+
+// Generate charts for both subscriber bills based on bill data
+
+const generateChartsForSubscriberBill = async (
+  subscriber,
+  latestBill,
+  previousBills
+) => {
+  console.log(`Generating Generation Data Bar Chart for ${subscriber.name}`);
+  const generationChart = generateGenerationDataChart(latestBill);
+  await saveChartToFile(generationChart, `${latestBill.id}_chart1`);
+
+  console.log(`Generating Cost Over Time chart for ${subscriber.name}`);
+  const costOverTimeChart = generateCostOverTimeChart(
+    previousBills,
+    latestBill
+  );
+  await saveChartToFile(costOverTimeChart, `${latestBill.id}_chart2`);
+
+  console.log(`Finished generating charts for ${subscriber.name}`);
+};
+
+// Generate PDF for subscriber's latest bill
+const generatePdfForSubscriber = async (
+  subscriberId,
+  freshBillGeneration = false
+) => {
+  // Refresh information about subscriber
+  const {
+    subscriber,
+    solarProject,
+    latestBill,
+    previousBills
+  } = await getLatestSubscriberData(subscriberId);
+
+  // Generate Charts (images saved locally)
+  await generateChartsForSubscriberBill(subscriber, latestBill, previousBills);
+
+  // Generate PDF for latest bill
+  console.log(`Creating PDF for Bill# ${latestBill.id} ...`);
+
+  // Create a unique timestamp to ensure unique URL for pdf
+  const pdfStamp = new Date().getTime();
+  await ReactPDF.render(
+    <BillTemplate
+      subscriber={subscriber}
+      solarProject={solarProject}
+      subscriberBill={latestBill}
+      previousBills={previousBills}
+    />,
+    `./temp/${latestBill.id}_${pdfStamp}.pdf`
```

Comment:
i would make the 'temp' folder name a variable on Constants.js just so it's easier to modify/isn't hardcoded

Reply:
it's used in quite a lot of places (below), and I foresee a lot of pain needing to change them in all of those places

Reply:
eh, in what world would the name of the temp folder ever be changed tho?? It serves a pretty standard purpose, does not seem worth

Reply:
as it is right to have such a folder hardcoded in different places creates a loosely coupled arrangement that is prone to errors/bugs in the future. Not only does declaring this in a constant make this easier to change, but it forestalls errors like that